### PR TITLE
185425763 outgoing referrals

### PR DESCRIPTION
**Pull request by ttoomey**
https://www.pivotaltracker.com/story/show/185425763
https://github.com/greenriver/hmis-frontend/pull/343

### CODE REVIEW

File: drivers/hmis/app/graphql/types/forms/pick_list_option.rb [hunk 1]
```diff
@@ -84,11 +88,33 @@ def self.options_for_type(pick_list_type, user:, relation_id: nil)
...
+        end
       when 'CLIENT_ENROLLMENTS'
         client = Hmis::Hud::Client.viewable_by(user).find_by(id: relation_id)
         return [] unless client.present?
 
-        client.enrollments.sort_by_option(:most_recent).map do |enrollment|
+        enrollments = client.enrollments
+          .preload(:project, :exit)
+
+        enrollments.sort_by_option(:most_recent).map do |enrollment|
```
Comment (ttoomey) [inline drivers/hmis/app/graphql/types/forms/pick_list_option.rb:118]: n+1 fix

File: drivers/hmis/app/graphql/mutations/ac_hmis/create_outgoing_referral_posting.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        HmisExternalApis::AcHmis::ReferralHouseholdMember.new(
+          relationship_to_hoh: member.relationship_to_hoh,
+          client_id: member.client.id
+        )
+      end
+
+      posting = referral.postings.build(
+        status: 'assigned_status',
+        project: project,
+        unit_type_id: input.unit_type_id,
```
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/ac_hmis/create_outgoing_referral_posting.rb:51]: I think we should error if the unit type is not available in the project. Something like: `project.units.unoccupied_on.pluck(:unit_type_id).uniq` should give the available unit types

File: drivers/hmis/app/graphql/mutations/ac_hmis/create_outgoing_referral_posting.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        data_source: enrollment.data_source,
+      )
+      posting.current_user = current_user
+
+      errors = HmisErrors::Errors.new
+      posting.transaction do
+        referral.save
+        errors.add_ar_errors(referral.errors.errors)
+        posting.save # context for validations
```
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/ac_hmis/create_outgoing_referral_posting.rb]: I'm not sure we need a context validation here, but wouldn't hurt I guess. We want to ensure enrollment_id is present and that status is assigned, and those things are both already verified in this mutation.

File: drivers/hmis/app/graphql/mutations/ac_hmis/create_outgoing_referral_posting.rb (hunk 2 above)
Comment (ttoomey) [reply drivers/hmis/app/graphql/mutations/ac_hmis/create_outgoing_referral_posting.rb:51]: added validation per our discussion

File: drivers/hmis/app/graphql/mutations/ac_hmis/create_outgoing_referral_posting.rb (hunk 3 above)
Comment (ttoomey) [reply drivers/hmis/app/graphql/mutations/ac_hmis/create_outgoing_referral_posting.rb]: added validation in the mutation

**Review by gigxz — APPROVED** [review]

**Pull request merged**